design/design-doc-finalizer-support.md

```diff
@@ -0,0 +1,332 @@
+resource.  This includes "local" Composite or Managed resources as well as "remote" resources that
+a Managed resource represents.
+
+## Out of Scope
+
+This design does not consider any changes to the Claim/Composite reference design that would
+allow __kubectl delete --cascade=foreground__ to work from the command line on a Claim resource.
+
+This design is not intended to block resource deletion requests to the Kubernetes API.
+
+This design does not address use of the Orphan propagation policy when deleting Claims or
+standalone Composite objects.
+
+## Design
+
+The intent of this design is to allow Crossplane objects to "behave" like native Kubernetes objects
+during delete processing.  A native Kubernetes object will remain in etcd, and all of it's
+content remains available to the user, while the object is marked deleted and waiting for
+finalizers to be removed by any controllers that are watching the resource.
+
+Adding _blockOwnerDeletion: true_ to the Controller ownerReference is consistent with how
+Kubernetes validates Controller references[block], and is required for Foreground Owner Deletion
+to work.
+
+Since Crossplane Composite resources represent collections of Composed objects, and Managed
+resources represent remote resources in other environments, a finalizer on the resource should
+preserve the existence of the related local/remote resources associated with the composite/managed
+resource.
+
+Given the restrictions on Claim/Composite references, in order to simulate Foreground Cascading
+Deletion on claims, a new attribute is added to the Claim.
+
+### API
+
+#### Claim: compositeDeletePolicy
+
+In order to indicate to the claim reconciler that it should use a specific Propagation Policy
+when deleting the associated Composite, a compositeDeletePolicy attribute is added to the Claim
+spec.  This attribute has possible values of Background (default) and Foreground.  The value
+"Orphan" is intentionally not supported as there is no requirement to leave the Composite
+resource in place (and under reconciliation) while deleting the Claim.
+
+The Claim reconciler will use the value of compositeDeletePolicy in the Delete API call for
+the associated Composite.  If Background is used the Composite deletion will occur as it does
+today and all associated resources will be marked for deletion immediately and then reconciled.
+
+If the value Foreground is used then the Composite and all of it's composed resources will be
+deleted using Foreground Cascading Deletion and the delete process with start at the leaf nodes
+and proceed back to the composite.
+
+**Examples:**
+
+Delete the Composite resource using Background propagation policy (already existing case):
+
+```yaml
+spec:
+  compositeDeletePolicy: Background
+```
+
+Delete the Composite resource using Foreground propagation policy:
+
+```yaml
+spec:
+  compositeDeletePolicy: Foreground
+```
```

Comment:
I was thinking today - is it instead possible to infer that a claim was deleted with `propagationPolicy: Foreground` by looking for the presence of the `foregroundDeletion` finalizer on the claim? If it were this would seem like a more "transparent" implementation of cascading deletion than adding a new field (which may conflict with the policy supplied when deleting the claim).

I have a vague recollection that you looked into this already and the answer is "no", but I can't find where we discussed it.

Reply:
That was my initial plan but when I tested the changes, I placed a watch on the claim and found that the foregroundDeletion finalizer is placed on the claim and then immediately removed, presumably because there are no ownerReferences pointing at the claim, and by the time the reconciler retrieves the object the foregroundDeletion finalizer is gone.  So the reconciler has no way to know that the deletion was caused by foreground deletion.   I don't know of any other way to detect foreground cascading deletion has been requested.  It's too bad there isn't an annotation added with that information.

Reply:
Got it.

I'd love it if we could somehow detect foreground deletion of a claim without needing this separate policy, but I don't see any great options.